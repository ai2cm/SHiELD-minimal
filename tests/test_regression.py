import hashlib
import os
import subprocess
import typing

import fv3config
import pytest

from pathlib import Path


CONFIG_DIR = Path(__file__).parent
EXECUTABLE = Path("/SHiELD/SHiELD_build/Build/bin/SHiELD_nh.prod.64bit.gnu.x")


def get_config(filename):
    config_filename = CONFIG_DIR / filename
    with open(config_filename, "r") as f:
        return fv3config.load(f)


def checksum_file(path: Path) -> str:
    sum = hashlib.md5()
    BUFFER_SIZE = 1024 * 1024
    with open(path, "rb") as f:
        while True:
            buf = f.read(BUFFER_SIZE)
            if not buf:
                break
            sum.update(buf)
    return sum.hexdigest()


def get_rundir_netcdfs(rundir: Path) -> typing.List[Path]:
    restart_directory = rundir / "RESTART"
    diagnostics_files = sorted(rundir.glob("*.nc"))
    restart_files = sorted(restart_directory.glob("*.nc"))
    return diagnostics_files + restart_files    
        

def checksum_rundir_to_file(rundir: Path, file: Path):
    """checksum rundir storing output in file"""
    rundir_netcdfs = get_rundir_netcdfs(rundir)
    for path in rundir_netcdfs:
        print(path, checksum_file(path), file=file)


def _run_simulation(config: dict, rundir: Path, command: typing.List[str]):
    fv3config.write_run_directory(config, rundir)
    n_processes = fv3config.config.get_n_processes(config)
    command = ["mpirun", "-n", str(n_processes)] + command
    completed_process = subprocess.run(command, cwd=rundir, capture_output=True)
    if completed_process.returncode != 0:
        print("Tail of Stderr:")
        print(completed_process.stderr[-2000:].decode())
        print("Tail of Stdout:")
        print(completed_process.stdout[-2000:].decode())
        pytest.fail()
    return completed_process
    
        
def run_fortran_executable(config: dict, rundir: Path):
    command = [EXECUTABLE.absolute().as_posix()]
    return _run_simulation(config, rundir, command)    


def test_regression_fortran(tmp_path: Path, regtest):
    config = get_config("default.yml")
    rundir_fortran = tmp_path / "rundir-fortran"

    run_fortran_executable(config, rundir_fortran)
    checksum_rundir_to_file(rundir_fortran, file=regtest)
