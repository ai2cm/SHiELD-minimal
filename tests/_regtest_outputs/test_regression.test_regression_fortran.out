<pytest_tempdir>/test_regression_fortran0/rundir/atmos_dt_atmos.tile1.nc 4326450c58c9c9e06c74339f99a04ed4
<pytest_tempdir>/test_regression_fortran0/rundir/atmos_dt_atmos.tile2.nc ce0f5c6b97f22ebf00f557f75c3a344b
<pytest_tempdir>/test_regression_fortran0/rundir/atmos_dt_atmos.tile3.nc 7615481471284e3fc46fa0a24618a7f9
<pytest_tempdir>/test_regression_fortran0/rundir/atmos_dt_atmos.tile4.nc dfcaa8f6555b47ff77453d6e787b9ac7
<pytest_tempdir>/test_regression_fortran0/rundir/atmos_dt_atmos.tile5.nc a50c349aedf609ddc3d8057973e70243
<pytest_tempdir>/test_regression_fortran0/rundir/atmos_dt_atmos.tile6.nc 9272be44b2873c6a27d553796826caa9
<pytest_tempdir>/test_regression_fortran0/rundir/atmos_static.tile1.nc 0dc74e89e2bd6ee5e91ac55ed96f30bd
<pytest_tempdir>/test_regression_fortran0/rundir/atmos_static.tile2.nc c8e66f31407e4b360f40e8f2d6348cd7
<pytest_tempdir>/test_regression_fortran0/rundir/atmos_static.tile3.nc 8efd8a973ac687ed854b495f3a279bad
<pytest_tempdir>/test_regression_fortran0/rundir/atmos_static.tile4.nc 85a260fdcdf3748f47b6a5c530e145e2
<pytest_tempdir>/test_regression_fortran0/rundir/atmos_static.tile5.nc 822ba7caefdc00317d4058b9e8dd4fa5
<pytest_tempdir>/test_regression_fortran0/rundir/atmos_static.tile6.nc 664dfed0eaafb54f911b4acada7c9525
<pytest_tempdir>/test_regression_fortran0/rundir/sfc_dt_atmos.tile1.nc 060abb83dbb1836958b92a9b14f590df
<pytest_tempdir>/test_regression_fortran0/rundir/sfc_dt_atmos.tile2.nc 57750c29439be940d51ccff58ab80caa
<pytest_tempdir>/test_regression_fortran0/rundir/sfc_dt_atmos.tile3.nc c1f9c924b8cddc701cbe25631dfa262a
<pytest_tempdir>/test_regression_fortran0/rundir/sfc_dt_atmos.tile4.nc 635994c121711cb57da8292dd59a9a38
<pytest_tempdir>/test_regression_fortran0/rundir/sfc_dt_atmos.tile5.nc 4ce71c880aeecfcdf239c41b7a66847f
<pytest_tempdir>/test_regression_fortran0/rundir/sfc_dt_atmos.tile6.nc 1da29572413c264f526f988ea386cd4a
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_core.res.nc 18f136825b9be2c9daad27aff6e25375
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_core.res.tile1.nc cc0817811d11099012a586c2dad074a5
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_core.res.tile2.nc 0e40e413e369e1e1c4b131dc14897c2c
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_core.res.tile3.nc 93dcd4939f78da7fcb395a3586bf7c99
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_core.res.tile4.nc fcc09204020bb6a821834a18b3ce74fb
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_core.res.tile5.nc 60076ffcf25cbf0c53db55a3d0e5d0ae
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_core.res.tile6.nc 4927c9d9340d993f85bfd3db7d2ff86a
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_srf_wnd.res.tile1.nc 34a2241cc73fa07baa27b7ed7fae07aa
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_srf_wnd.res.tile2.nc e3159ba05783fd2000a8b0afbbd4fc02
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_srf_wnd.res.tile3.nc 2e1d5f422d73101c135d1ffa663583dd
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_srf_wnd.res.tile4.nc 84dfba367bc2e2fdd056cdeb018dec86
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_srf_wnd.res.tile5.nc b7d04050c72aea38288e6fe14958b088
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_srf_wnd.res.tile6.nc b466057b46cbfc037d6ac52785cbb0ea
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_tracer.res.tile1.nc 8cd98320e278cf5956e7015ca47018f4
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_tracer.res.tile2.nc 3bb17be7b26116922299c3d8d9edd8d3
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_tracer.res.tile3.nc fc0c26959ff5f16b456c608196141cec
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_tracer.res.tile4.nc 5c8a3a2d5935b29ecd6153d1b260a731
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_tracer.res.tile5.nc dbfd16d9d675d52c98025802ed93745a
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/fv_tracer.res.tile6.nc 96ea2df4662cc3c75df21bbbd1174559
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/phy_data.tile1.nc c907b002740b7090e7752fcfbc0365c8
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/phy_data.tile2.nc ebf06b07b304eeef39e5be80be91beaa
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/phy_data.tile3.nc fad4c687044b81103a2efaf5e7f17c76
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/phy_data.tile4.nc 14f25ea7fbd979ff91108a15096341c2
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/phy_data.tile5.nc b52a13593daa5322a4f726ef9e8a344c
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/phy_data.tile6.nc 8c1ddf1db8cb4ba7df8e1b3377db8721
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/sfc_data.tile1.nc 0a4277edf597338c08517c6732a619f9
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/sfc_data.tile2.nc 47366511dc98934481167dcb2c987b8b
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/sfc_data.tile3.nc b6cd12f07071866965e70fc05aea5bdb
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/sfc_data.tile4.nc dbae57da6b2aaafcca0dad0999b56cf6
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/sfc_data.tile5.nc ad06c9da4c0b2a96400c0bcdece1d9fe
<pytest_tempdir>/test_regression_fortran0/rundir/RESTART/sfc_data.tile6.nc 7482a30d6f3192404f1d5c01f9e45af3
